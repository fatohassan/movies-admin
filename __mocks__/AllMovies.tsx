import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { renderHook } from "@testing-library/react";
import React from "react";
import  MoviesList  from "../src/pages/admin/movies/MoviesList";
import { LOAD_MOVIES } from "../src/pages/admin/movieQueries/moviesQueries";

export const AllMovies = {
    getMoviesList: [
    {
      id: "66715c4f88c42067190e070f",
      original_title: "Inside Out 2",
      release_date: "2024-06-11",
      vote_average: "7.761",
      backdrop_path: "/coATv42PoiLqAFKStJiMZs2r6Zb.jpg",
      image: null,
      overview:
        "Teenager Riley's mind headquarters is undergoing a sudden demolition to make room for something entirely unexpected: new Emotions! Joy, Sadness, Anger, Fear and Disgust, who’ve long been running a successful operation by all accounts, aren’t sure how to feel when Anxiety shows up. And it looks like she’s not alone.",
    },
  ],
};

// successful mock of getMovieList
export const successfulMoviesMock: MockedResponse[]= [
    {
        request: {
            query: LOAD_MOVIES
        },
        result: {
            data: AllMovies
        }
    }
] 

export function getAllMoviesWrapper(mockData: MockedResponse[] = []) {
    const wrapper =({children}: React.PropsWithChildren) => (
        <MockedProvider mocks={mockData} addTypename={false}>
            {children}
        </MockedProvider>
    );
    const {result}:any = renderHook(() => MoviesList(), {wrapper})
    return {
        result
    }
}
