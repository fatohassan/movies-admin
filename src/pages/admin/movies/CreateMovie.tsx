import { Link, useNavigate } from "react-router-dom";
import { useState } from "react";
import { useMutation } from "@apollo/client";
import { CREATE_MOVIE_MUTATION } from "../movieQueries/moviesMutations";

export default function CreateMovie() {
  const [form, setForm] = useState<any>({
    original_title: "",
    release_date: "",
    vote_average: "",
    backdrop_path: "",
    overview: "",
  });

  const [image, setImage] = useState("");
  const [createMovie] = useMutation(CREATE_MOVIE_MUTATION);
  const navigate = useNavigate();


  const uploadImage = async (e) => {
    const file = e.target.files[0];
    const base64 = await convertBase64(file);
    setImage(base64 as string);
  };

  const convertBase64 = (file) => {
    return new Promise((resolve, reject) => {
      const fileReader = new FileReader();
      fileReader.readAsDataURL(file);
      fileReader.onload = () => {
        resolve(fileReader.result);

        fileReader.onerror = (error) => {
          reject(error);
        };
      };
    });
  };

  const handleSubmit = async (event: React.FormEvent<any>) => {
    try {
      await createMovie({
        variables: {
          movieInput: {
            original_title: form.original_title,
            release_date: form.release_date,
            vote_average: form.vote_average,
            backdrop_path: form.backdrop_path,
            overview: form.overview,
            image: image,
          },
        },
      });
      
      setForm({
        original_title: "",
        release_date: "",
        vote_average: "",
        backdrop_path: "",
        overview: "",
      });
      navigate("/admin/movies/");
    } catch (error) {
      console.log(error);
     
    }
  };

  return (
    <div className="container my-4">
      <div className="row">
        <div className="col-md-8 mx-auto rounded border p-4">
          <h2 className="text-center mb-5">Create Movie</h2>

          <form onSubmit={handleSubmit}>
            <div className="row mb-3">
              <label className="col-sm-4 col-form-label">Title</label>
              <div className="col-sm-8">
                <input
                  className="form-control"
                  id="original_title"
                  value={form.original_title}
                  name="original_title"
                  onChange={(event: any) =>
                    setForm({
                      ...form,
                      original_title: event.currentTarget.value,
                    })
                  }
                />
                <span className="text-danger"></span>
              </div>
            </div>
            <div className="row mb-3">
              <label className="col-sm-4 col-form-label">Rating</label>
              <div className="col-sm-8">
                <input
                  className="form-control"
                  id="vote_average"
                  value={form.vote_average}
                  name="vote_average"
                  onChange={(event: any) =>
                    setForm({
                      ...form,
                      vote_average: event.currentTarget.value,
                    })
                  }
                  type="number"
                />
                <span className="text-danger"></span>
              </div>
            </div>
            <div className="row mb-3">
              <label className="col-sm-4 col-form-label">Release Date</label>
              <div className="col-sm-8">
                <input
                  className="form-control"
                  id="release_date"
                  value={form.release_date}
                  name="release_date"
                  onChange={(event: any) =>
                    setForm({
                      ...form,
                      release_date: event.currentTarget.value,
                    })
                  }
                />
                <span className="text-danger"></span>
              </div>
            </div>
            <div className="row mb-3">
              <label className="col-sm-4 col-form-label">Description</label>
              <div className="col-sm-8">
                <textarea
                  className="form-control"
                  id="overview"
                  value={form.overview}
                  name="overview"
                  onChange={(event: any) =>
                    setForm({ ...form, overview: event.currentTarget.value })
                  }
                  rows={3}
                />
                <span className="text-danger"></span>
              </div>
            </div>
            <div className="row mb-3">
              <label className="col-sm-4 col-form-label">Image</label>
              <div className="col-sm-8">
                <input
                  className="form-control"
                  id="image"
                  name="image"
                  onChange={(event) => {
                    uploadImage(event);
                  }}
                  type="file"
                />
                <span className="text-danger"></span>
              </div>
            </div>

            <div className="row">
              <div className="offset-sm-4 col-sm-4 d-grid">
                <button type="submit" className="btn btn-primary">
                  Submit
                </button>
              </div>
              <div className="col-sm-4 d-grid">
                <Link
                  className="btn btn-secondary"
                  to="/admin/movies"
                  role="button"
                >
                  Cancel
                </Link>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  );
}
