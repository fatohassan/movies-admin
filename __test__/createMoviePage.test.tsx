import { cleanup, fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import CreateMovie from "../src/pages/admin/movies/CreateMovie";
import React from "react";
import renderer from "react-test-renderer";
import { MockedProvider } from "@apollo/client/testing";
import { BrowserRouter } from "react-router-dom";
import { createMovieMutationsMock } from "../__mocks__/AllMovies";

const renderCreateMovie = () => {
  return (
    <BrowserRouter>
      <MockedProvider mocks={createMovieMutationsMock} addTypename={false}>
        <CreateMovie />
      </MockedProvider>
    </BrowserRouter>
  );
};

describe("Create Movie rendering navigating", () => {
  it("should render page correctly", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <MockedProvider mocks={createMovieMutationsMock} addTypename={false}>
            <CreateMovie />
          </MockedProvider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should display contents", () => {
    render(renderCreateMovie());
    const headingElement = screen.getByRole("heading", {
      name: /Create Movie/i,
    });
    expect(headingElement).toBeInTheDocument();
  });

  it("should display form contents/inputs/buttons", async () => {
    render(renderCreateMovie());
    expect(screen.getByTestId("form")).toBeInTheDocument();
    expect(screen.getByText("Title")).toBeInTheDocument();
    expect(screen.getByText("Rating")).toBeInTheDocument();
    expect(screen.getByText("Image")).toBeInTheDocument();
    expect(screen.getByText("Description")).toBeInTheDocument();
    expect(screen.getByText("Release Date")).toBeInTheDocument();
    expect(screen.getByLabelText("title")).toBeInTheDocument();
    expect(screen.getByLabelText("rating")).toBeInTheDocument();
    expect(screen.getByLabelText("date")).toBeInTheDocument();
    expect(screen.getByLabelText("image")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "Submit" })).toBeTruthy();
    expect(screen.getByLabelText("link")).toHaveTextContent("Cancel");
  });

  it("should be able to type in title inuput", () => {
    render(renderCreateMovie());
    const inputTitle = screen.getByLabelText("title");
    expect(
      fireEvent.change(inputTitle, { EventTarget: { value: "Inside Out 2" } })
    ).toBeTruthy();
  });

  it("should be able to type in rating inuput", () => {
    render(renderCreateMovie());
    const inputRating = screen.getByLabelText("rating");
    expect(
      fireEvent.change(inputRating, { EventTarget: { value: "1" } })
    ).toBeTruthy();
  });

  it("should be able to type in description inuput", () => {
    render(renderCreateMovie());
    const inputOverview = screen.getByLabelText("overview");
    fireEvent.change(inputOverview, { EventTarget: { value: "Good Movie" } });
    expect(inputOverview).toBeVisible();
  });

  it("should be able to type in description inuput", () => {
    render(renderCreateMovie());
    const inputReleaseDate = screen.getByLabelText("date");
    expect(
      fireEvent.change(inputReleaseDate, {
        EventTarget: { value: "01-02-2024" },
      })
    ).not.toBeFalsy();
  });

  it("should be able to type in image inuput", () => {
    render(renderCreateMovie());
    const inputImage = screen.getByLabelText("image");
    expect(inputImage).toBeInTheDocument();
  });
});
