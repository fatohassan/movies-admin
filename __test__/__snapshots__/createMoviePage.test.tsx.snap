// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Movie rendering navigating should render links correctly 1`] = `
<div
  className="container my-4"
>
  <div
    className="row"
  >
    <div
      className="col-md-8 mx-auto rounded border p-4"
    >
      <h2
        className="text-center mb-5"
      >
        Create Movie
      </h2>
      <form
        data-testid="form"
        onSubmit={[Function]}
      >
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Title
          </label>
          <div
            className="col-sm-8"
          >
            <input
              aria-label="title"
              className="form-control"
              id="original_title"
              name="original_title"
              onChange={[Function]}
              value=""
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Rating
          </label>
          <div
            className="col-sm-8"
          >
            <input
              aria-label="rating"
              className="form-control"
              id="vote_average"
              name="vote_average"
              onChange={[Function]}
              type="number"
              value=""
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Release Date
          </label>
          <div
            className="col-sm-8"
          >
            <input
              aria-label="date"
              className="form-control"
              id="release_date"
              name="release_date"
              onChange={[Function]}
              value=""
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Description
          </label>
          <div
            className="col-sm-8"
          >
            <textarea
              aria-label="overview"
              className="form-control"
              id="overview"
              name="overview"
              onChange={[Function]}
              rows={3}
              value=""
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Image
          </label>
          <div
            className="col-sm-8"
          >
            <input
              aria-label="image"
              className="form-control"
              id="image"
              name="image"
              onChange={[Function]}
              type="file"
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="offset-sm-4 col-sm-4 d-grid"
          >
            <button
              className="btn btn-primary"
              type="submit"
            >
              Submit
            </button>
          </div>
          <div
            className="col-sm-4 d-grid"
          >
            <a
              aria-label="link"
              className="btn btn-secondary"
              href="/admin/movies"
              onClick={[Function]}
              role="button"
            >
              Cancel
            </a>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`Create Movie rendering navigating should render page correctly 1`] = `
<div
  className="container my-4"
>
  <div
    className="row"
  >
    <div
      className="col-md-8 mx-auto rounded border p-4"
    >
      <h2
        className="text-center mb-5"
      >
        Create Movie
      </h2>
      <form
        data-testid="form"
        onSubmit={[Function]}
      >
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Title
          </label>
          <div
            className="col-sm-8"
          >
            <input
              aria-label="title"
              className="form-control"
              id="original_title"
              name="original_title"
              onChange={[Function]}
              value=""
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Rating
          </label>
          <div
            className="col-sm-8"
          >
            <input
              aria-label="rating"
              className="form-control"
              id="vote_average"
              name="vote_average"
              onChange={[Function]}
              type="number"
              value=""
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Release Date
          </label>
          <div
            className="col-sm-8"
          >
            <input
              aria-label="date"
              className="form-control"
              id="release_date"
              name="release_date"
              onChange={[Function]}
              value=""
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Description
          </label>
          <div
            className="col-sm-8"
          >
            <textarea
              aria-label="overview"
              className="form-control"
              id="overview"
              name="overview"
              onChange={[Function]}
              rows={3}
              value=""
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Image
          </label>
          <div
            className="col-sm-8"
          >
            <input
              aria-label="image"
              className="form-control"
              id="image"
              name="image"
              onChange={[Function]}
              type="file"
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="offset-sm-4 col-sm-4 d-grid"
          >
            <button
              className="btn btn-primary"
              type="submit"
            >
              Submit
            </button>
          </div>
          <div
            className="col-sm-4 d-grid"
          >
            <a
              aria-label="link"
              className="btn btn-secondary"
              href="/admin/movies"
              onClick={[Function]}
              role="button"
            >
              Cancel
            </a>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
